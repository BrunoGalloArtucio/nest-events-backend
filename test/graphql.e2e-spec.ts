import * as request from 'supertest';
import { Test, TestingModule } from '@nestjs/testing';
import { DataSource } from 'typeorm';
import { AppModule } from '../src/app.module';
import {
  loadFixtures as loadFixturesBase,
  tokenForUser as tokenForUserBase,
} from './utils';
import { INestApplication, ValidationPipe } from '@nestjs/common';
import { User } from 'src/auth/user.entity';
import { ExecutionResult } from 'graphql';
import { PaginatedTeachers } from 'src/school/teacher.entity';

let app: INestApplication;
let mod: TestingModule;
let dataSource: DataSource;

const loadFixtures = async (sqlFileName: string) =>
  loadFixturesBase(dataSource, sqlFileName);

const tokenForUser = (
  user: Partial<User> = {
    id: 1,
    username: 'e2e-test',
  },
): string => tokenForUserBase(app, user);

describe('GraphQL (e2e)', () => {
  beforeEach(async () => {
    mod = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = mod.createNestApplication();
    app.useGlobalPipes(new ValidationPipe());

    await app.init();

    dataSource = app.get(DataSource);
  });

  afterEach(async () => {
    await app.close();
  });

  describe('teachers query', () => {
    test('should return empty paginated result', async () => {
      const response = await request(app.getHttpServer())
        .post('/graphql')
        .send({
          query: `
            query {
              teachers {
                total
                data {
                  id
                  name
                  gender
                  subjects{
                    id
                    name
                  }
                }
              }
            }`,
        });

      const body = response.body as ExecutionResult<{
        teachers: PaginatedTeachers;
      }>;
      expect(body.data.teachers.data).toEqual([]);
      expect(body.data.teachers.total).toBe(0);
    });

    test('should return paginated result with teachers', async () => {
      await loadFixtures('2-teachers-1-user.sql');

      const response = await request(app.getHttpServer())
        .post('/graphql')
        .send({
          query: `
            query {
              teachers {
                total
                data {
                  id
                  name
                  gender
                  subjects{
                    id
                    name
                  }
                }
              }
            }`,
        });

      const body = response.body as ExecutionResult<{
        teachers: PaginatedTeachers;
      }>;
      expect(body.data.teachers.total).toBe(2);
      expect(body).toMatchSnapshot();
    });
  });
});
