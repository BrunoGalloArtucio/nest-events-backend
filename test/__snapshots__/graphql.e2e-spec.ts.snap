// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL (e2e) teachers query should return paginated result with teachers 1`] = `
{
  "data": {
    "teachers": {
      "data": [
        {
          "gender": "Female",
          "id": 2,
          "name": "Teacher 2",
          "subjects": [],
        },
        {
          "gender": "Male",
          "id": 1,
          "name": "Teacher 1",
          "subjects": [],
        },
      ],
      "total": 2,
    },
  },
}
`;
